# Employee Attrition - Data Audit

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/Employee_Attrition_Prediction.csv")

print("Shape:", df.shape)

Shape: (1470, 35)


In [4]:
df.sample(5, random_state=42)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1041,28,No,Travel_Rarely,866,Sales,5,3,Medical,1,1469,...,4,80,0,6,4,3,5,4,1,3
184,53,No,Travel_Rarely,1084,Research & Development,13,2,Medical,1,250,...,3,80,2,5,3,3,4,2,1,3
1222,24,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,...,3,80,1,1,2,3,1,0,0,0
67,45,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,...,3,80,1,25,2,3,1,0,0,0
220,36,No,Travel_Rarely,1396,Research & Development,5,2,Life Sciences,1,304,...,4,80,0,16,3,4,13,11,3,7


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [8]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [11]:
print(df["Attrition"].value_counts())

print(df["Attrition"].value_counts(normalize=True) * 100)

Attrition
No     1233
Yes     237
Name: count, dtype: int64
Attrition
No     83.877551
Yes    16.122449
Name: proportion, dtype: float64


In [12]:
nunique = df.nunique()

nunique[nunique == 1]

EmployeeCount    1
Over18           1
StandardHours    1
dtype: int64

In [13]:
df.nunique().sort_values(ascending=False).head(10)

EmployeeNumber        1470
MonthlyRate           1427
MonthlyIncome         1349
DailyRate              886
HourlyRate              71
Age                     43
TotalWorkingYears       40
YearsAtCompany          37
DistanceFromHome        29
YearsInCurrentRole      19
dtype: int64

In [15]:
df["EmployeeNumber"].nunique()

1470

In [20]:
for column in df.columns:
    print(f"\n{column}")
    print("Unique values:", df[column].nunique())
    print("data type:", df[column].dtype)


Age
Unique values: 43
data type: int64

Attrition
Unique values: 2
data type: object

BusinessTravel
Unique values: 3
data type: object

DailyRate
Unique values: 886
data type: int64

Department
Unique values: 3
data type: object

DistanceFromHome
Unique values: 29
data type: int64

Education
Unique values: 5
data type: int64

EducationField
Unique values: 6
data type: object

EmployeeCount
Unique values: 1
data type: int64

EmployeeNumber
Unique values: 1470
data type: int64

EnvironmentSatisfaction
Unique values: 4
data type: int64

Gender
Unique values: 2
data type: object

HourlyRate
Unique values: 71
data type: int64

JobInvolvement
Unique values: 4
data type: int64

JobLevel
Unique values: 5
data type: int64

JobRole
Unique values: 9
data type: object

JobSatisfaction
Unique values: 4
data type: int64

MaritalStatus
Unique values: 3
data type: object

MonthlyIncome
Unique values: 1349
data type: int64

MonthlyRate
Unique values: 1427
data type: int64

NumCompaniesWorked
Unique v

In [21]:
df.groupby("Attrition")["Age"].describe()

,count,mean,std,min,25%,50%,75%,max
Attrition,,,,,,,,
No,1233.0,37.561233,8.88836,18.0,31.0,36.0,43.0,60.0
Yes,237.0,33.607595,9.68935,18.0,28.0,32.0,39.0,58.0


In [22]:
df.groupby("Attrition")["MonthlyIncome"].describe()

,count,mean,std,min,25%,50%,75%,max
Attrition,,,,,,,,
No,1233.0,6832.739659,4818.208001,1051.0,3211.0,5204.0,8834.0,19999.0
Yes,237.0,4787.092827,3640.210367,1009.0,2373.0,3202.0,5916.0,19859.0


In [23]:
pd.crosstab(
    df["OverTime"],
    df["Attrition"],
    normalize="index"
) * 100

Attrition,No,Yes
OverTime,,
No,89.563567,10.436433
Yes,69.471154,30.528846


In [24]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,1470.0,36.923810,9.135373,18.0,30.00,36.0,43.00,60.0
DailyRate,1470.0,802.485714,403.509100,102.0,465.00,802.0,1157.00,1499.0
DistanceFromHome,1470.0,9.192517,8.106864,1.0,2.00,7.0,14.00,29.0
Education,1470.0,2.912925,1.024165,1.0,2.00,3.0,4.00,5.0
EmployeeCount,1470.0,1.000000,0.000000,1.0,1.00,1.0,1.00,1.0
EmployeeNumber,1470.0,1024.865306,602.024335,1.0,491.25,1020.5,1555.75,2068.0
EnvironmentSatisfaction,1470.0,2.721769,1.093082,1.0,2.00,3.0,4.00,4.0
HourlyRate,1470.0,65.891156,20.329428,30.0,48.00,66.0,83.75,100.0
JobInvolvement,1470.0,2.729932,0.711561,1.0,2.00,3.0,3.00,4.0
JobLevel,1470.0,2.063946,1.106940,1.0,1.00,2.0,3.00,5.0


In [25]:
df.select_dtypes(include="object").describe().T

,count,unique,top,freq
Attrition,1470,2,No,1233
BusinessTravel,1470,3,Travel_Rarely,1043
Department,1470,3,Research & Development,961
EducationField,1470,6,Life Sciences,606
Gender,1470,2,Male,882
JobRole,1470,9,Sales Executive,326
MaritalStatus,1470,3,Married,673
Over18,1470,1,Y,1470
OverTime,1470,2,No,1054


## Audit Summary

### Dataset
- 1,470 employee records
- 35 columns
- No missing values
- No duplicate rows

### Target
- `Attrition` is the target variable
- 83.88% No
- 16.12% Yes
- The target is imbalanced

### Columns removed
- `EmployeeNumber` — identifier
- `EmployeeCount` — constant
- `Over18` — constant
- `StandardHours` — constant

### Feature groups
- Numerical features
- Ordinal features
- Nominal categorical features

### Initial observations
- Employees who left are younger on average.
- Employees who left have lower average and median monthly income.
- Employees working overtime have substantially higher observed attrition.
- These are associations, not causal conclusions.

### Next step
Perform exploratory data analysis and investigate feature distributions
and relationships with attrition.